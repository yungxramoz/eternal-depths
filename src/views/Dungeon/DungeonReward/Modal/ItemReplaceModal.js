import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../../components/atoms/RpgButton/RpgButton'
import Modal from '../../../../components/organisms/Modal/Modal'
import { characterEquipItem } from '../../../../store/game/gameSlice'
import ItemDetails from '../../../../components/molecules/ItemDetails/ItemDetails'
import RpgSeparator from '../../../../components/templates/RpgSeperator/RpgSeparator'
import { getItemStats } from '../../../../utils/item-generator'

const ItemReplaceModal = ({ newItem, onClose, onRewardSelected }) => {
  const dispatch = useDispatch()
  const items = useSelector((state) => state.game.character.current.items)
  const currentItem = items[newItem.type]

  const itemStats = (item) => getItemStats(item)

  const handleReplace = () => {
    dispatch(characterEquipItem(newItem))
    onRewardSelected()
  }

  return (
    <Modal
      title="Replace Item"
      onDismiss={onClose}
      content={
        <div className="item-replace-modal">
          <p>Do you want to replace the current item:</p>
          <ItemDetails
            title={currentItem.name}
            subtitle={currentItem.rarity}
            description={itemStats(currentItem)}
            icon={currentItem.icon}
          />
          <RpgSeparator />
          <p>New item:</p>
          <ItemDetails
            title={newItem.name}
            subtitle={newItem.rarity}
            description={itemStats(newItem)}
            icon={newItem.icon}
          />
        </div>
      }
      footerContent={<RpgButton text="Replace" onClick={handleReplace} />}
    />
  )
}

export default ItemReplaceModal
