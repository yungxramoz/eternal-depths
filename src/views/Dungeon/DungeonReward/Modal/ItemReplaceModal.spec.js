import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../../../store/game/gameSlice'
import ItemReplaceModal from './ItemReplaceModal'

describe('ItemReplaceModal', () => {
  let store
  let onRewardSelected
  const item = {
    type: 'helmet',
    icon: 'test-icon',
    name: 'Test Item',
    rarity: 'Common',
    stats: { strength: 10 },
  }
  const preloadedState = {
    game: {
      character: {
        current: {
          items: {
            helmet: item,
          },
        },
      },
    },
  }

  beforeEach(() => {
    store = configureStore({ reducer: { game: gameReducer }, preloadedState })
    onRewardSelected = jest.fn()
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <ItemReplaceModal
          newItem={item}
          onClose={() => {}}
          onRewardSelected={onRewardSelected}
        />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('ItemReplaceModal')
  })

  it('dispatches characterEquipItem action and calls onRewardSelected when Replace button is clicked', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <ItemReplaceModal
          newItem={item}
          onClose={() => {}}
          onRewardSelected={onRewardSelected}
        />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Replace'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/characterEquipItem' }),
    )
    expect(onRewardSelected).toHaveBeenCalled()
  })
})
