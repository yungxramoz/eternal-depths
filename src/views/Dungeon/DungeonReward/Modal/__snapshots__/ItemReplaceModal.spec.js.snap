// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemReplaceModal renders correctly: ItemReplaceModal 1`] = `
<body>
  <div>
    <div
      class="modal-container"
    >
      <div
        class="rpgui-container container framed framed-golden modal"
      >
        <div
          class="modal-header"
        >
          <button
            class="icon-button large dismiss-button"
          >
            <svg
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 5h2v2H5V5zm4 4H7V7h2v2zm2 2H9V9h2v2zm2 0h-2v2H9v2H7v2H5v2h2v-2h2v-2h2v-2h2v2h2v2h2v2h2v-2h-2v-2h-2v-2h-2v-2zm2-2v2h-2V9h2zm2-2v2h-2V7h2zm0 0V5h2v2h-2z"
                fill="currentColor"
              />
            </svg>
          </button>
          <h2>
            Replace Item
          </h2>
          <hr
            class="separator"
          />
        </div>
        <div
          class="modal-body"
        >
          <div
            class="item-replace-modal"
          >
            <p>
              Do you want to replace the current item:
            </p>
            <div
              class="item-container"
            >
              <div
                class="rpgui-icon test-icon item-icon"
              />
              <div
                class="item-content"
              >
                <p
                  class="item-title"
                >
                  Test Item
                </p>
                <p
                  class="item-subtitle"
                >
                  <i>
                    Common
                  </i>
                </p>
                <p
                  class="item-description"
                >
                  STR:10
                </p>
              </div>
            </div>
            <hr
              class="separator golden"
            />
            <p>
              New item:
            </p>
            <div
              class="item-container"
            >
              <div
                class="rpgui-icon test-icon item-icon"
              />
              <div
                class="item-content"
              >
                <p
                  class="item-title"
                >
                  Test Item
                </p>
                <p
                  class="item-subtitle"
                >
                  <i>
                    Common
                  </i>
                </p>
                <p
                  class="item-description"
                >
                  STR:10
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <hr
            class="separator"
          />
          <button
            class="rpgui-button btn"
          >
            Replace
          </button>
        </div>
      </div>
      <div
        class="modal-background"
      />
    </div>
  </div>
</body>
`;
